File: src/memorygraph/utils/pagination.py
"""
Pagination utilities for working with memories across different backends.

Provides reusable pagination helpers that work with both MemoryDatabase
and SQLiteMemoryDatabase interfaces.
"""

import logging
from typing import AsyncIterator, List, Callable, Optional

from ..models import Memory, SearchQuery

logger = logging.getLogger(__name__)


async def paginate_memories(
    db,  # MemoryDatabase or SQLiteMemoryDatabase
    batch_size: int = 1000,
    progress_callback: Optional[Callable[[int], None]] = None
) -> AsyncIterator[List[Memory]]:
    """
    Async generator that yields batches of memories.

    Works with both MemoryDatabase and SQLiteMemoryDatabase by using
    the search_memories interface with pagination support.

    Args:
        db: Database instance (any backend)
        batch_size: Number of memories to fetch per batch
        progress_callback: Optional callback(total_yielded) for progress reporting

    Yields:
        Batches of Memory objects

    Example:
        async for batch in paginate_memories(db, batch_size=500):
            for memory in batch:
                print(memory.title)
    """
    offset = 0
    total_yielded = 0

    while True:
        query = SearchQuery(
            query="",  # Empty query matches all memories
            limit=batch_size,
            offset=offset,
            match_mode="any"
        )

        # Use paginated search if available, fallback to regular search
        if hasattr(db, 'search_memories_paginated'):
            result = await db.search_memories_paginated(query)
            memories = result.results
            has_more = result.has_more
        else:
            # Fallback for backends without pagination
            memories = await db.search_memories(query)
            has_more = len(memories) >= batch_size

        if memories:
            yield memories
            total_yielded += len(memories)

            if progress_callback:
                progress_callback(total_yielded)

        if not has_more or not memories:
            break

        offset += batch_size


async def count_memories(db) -> int:
    """
    Count total memories in database using efficient method.

    Args:
        db: Database instance (any backend)

    Returns:
        Total number of memories
    """
    query = SearchQuery(query="", limit=1, offset=0, match_mode="any")

    # Use paginated search for efficient counting if available
    if hasattr(db, 'search_memories_paginated'):
        result = await db.search_memories_paginated(query)
        return result.total_count
    else:
        # Fallback: count manually by iterating
        count = 0
        async for batch in paginate_memories(db, batch_size=1000):
            count += len(batch)
        return count


async def count_relationships(db) -> int:
    """
    Count total relationships in database.

    Args:
        db: Database instance (any backend)

    Returns:
        Total number of relationships (deduplicated)
    """
    count = 0
    seen_relationships = set()

    async for batch in paginate_memories(db, batch_size=1000):
        for memory in batch:
            try:
                related = await db.get_related_memories(memory.id, max_depth=1)
                for _, relationship in related:
                    # Use tuple as key for deduplication
                    key = (
                        relationship.from_memory_id,
                        relationship.to_memory_id,
                        relationship.type.value
                    )
                    if key not in seen_relationships:
                        seen_relationships.add(key)
                        count += 1
            except Exception as e:
                logger.warning(f"Failed to count relationships for memory {memory.id}: {e}")
                continue

    return count


async def get_all_memories(db) -> List[Memory]:
    """
    Get all memories from database.

    Args:
        db: Database instance (any backend)

    Returns:
        List of all memories

    Warning:
        This loads all memories into memory. For large databases,
        prefer using paginate_memories() to process in batches.
    """
    all_memories = []
    async for batch in paginate_memories(db, batch_size=1000):
        all_memories.extend(batch)
    return all_memories

File: ts/src/utils/pagination.ts
/**
 * Pagination utilities for working with memories across different backends.
 */

import type { Memory, SearchQuery } from "../models.js";

export interface DBLike {
  searchMemories(query: SearchQuery): Promise<Memory[]>;
  searchMemoriesPaginated?(query: SearchQuery): Promise<import("../models.js").PaginatedResult>;
}

export async function* paginateMemories(
  db: DBLike,
  batchSize = 1000,
  progressCallback?: (total: number) => void
): AsyncGenerator<Memory[]> {
  let offset = 0;
  let totalYielded = 0;

  while (true) {
    const query: SearchQuery = {
      query: undefined,
      terms: [],
      memory_types: [],
      tags: [],
      project_path: undefined,
      languages: [],
      frameworks: [],
      min_importance: undefined,
      min_confidence: undefined,
      min_effectiveness: undefined,
      created_after: undefined,
      created_before: undefined,
      limit: batchSize,
      offset,
      include_relationships: true,
      search_tolerance: "normal",
      match_mode: "any",
      relationship_filter: undefined,
    };

    let memories: Memory[];
    let hasMore: boolean;

    if (db.searchMemoriesPaginated) {
      const result = await db.searchMemoriesPaginated(query);
      memories = result.results;
      hasMore = result.has_more;
    } else {
      memories = await db.searchMemories(query);
      hasMore = memories.length >= batchSize;
    }

    if (memories.length > 0) {
      yield memories;
      totalYielded += memories.length;
      progressCallback?.(totalYielded);
    }

    if (!hasMore || memories.length === 0) break;
    offset += batchSize;
  }
}

export async function countMemories(db: DBLike): Promise<number> {
  const query: SearchQuery = {
    query: undefined,
    terms: [],
    memory_types: [],
    tags: [],
    project_path: undefined,
    languages: [],
    frameworks: [],
    min_importance: undefined,
    min_confidence: undefined,
    min_effectiveness: undefined,
    created_after: undefined,
    created_before: undefined,
    limit: 1,
    offset: 0,
    include_relationships: true,
    search_tolerance: "normal",
    match_mode: "any",
    relationship_filter: undefined,
  };

  if (db.searchMemoriesPaginated) {
    const result = await db.searchMemoriesPaginated(query);
    return result.total_count;
  }

  let count = 0;
  for await (const batch of paginateMemories(db, 1000)) {
    count += batch.length;
  }
  return count;
}

export async function getAllMemories(db: DBLike): Promise<Memory[]> {
  const all: Memory[] = [];
  for await (const batch of paginateMemories(db, 1000)) {
    all.push(...batch);
  }
  return all;
}
